shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Value.hs

```diff
@@ -0,0 +1,407 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Shelley.Spec.Ledger.Value
+ where
+
+import           Cardano.Binary (ToCBOR, FromCBOR, toCBOR, fromCBOR, encodeListLen,
+                  decodeWord)
+import           Cardano.Prelude (NoUnexpectedThunks(..))
+import           Data.Coerce (coerce)
+
+import           Shelley.Spec.Ledger.BaseTypes (invalidKey)
+import           Shelley.Spec.Ledger.Coin (Coin (..))
+import           GHC.Generics (Generic)
+import           Data.Word (Word8)
+import           Cardano.Crypto.Hash (Hash, ShortHash)
+--import           Data.Group (Group (..))
+--import           Data.Semigroup
+--import           Data.Monoid
+-- import           Algebra.Module
+import           Data.Map.Strict (Map, elems, empty, toList, filterWithKey, keys,
+                 toList, singleton, lookup, insert, unionWith)
+import           Shelley.Spec.Ledger.Crypto
+import           Data.ByteString.Char8 (ByteString, pack) -- TODO is this the right Bytestring
+import           Shelley.Spec.Ledger.Scripts
+
+
+
+
+-- | Quantity
+newtype Quantity = Quantity Integer
+  deriving (Show, Eq, Generic, ToCBOR, FromCBOR, Ord, Integral, Num, Real, Enum, NoUnexpectedThunks)
+
+-- | Asset ID
+newtype AssetID = AssetID ByteString
+  deriving (Show, Eq, ToCBOR, FromCBOR, Ord, NoUnexpectedThunks)
+
+assetID :: AssetID -> ByteString
+assetID (AssetID aid) = aid
+
+-- | Policy ID
+newtype PolicyID crypto = PolicyID (ScriptHash crypto)
+  deriving (Show, Eq, ToCBOR, FromCBOR, Ord, NoUnexpectedThunks)
+
+policyID :: PolicyID crypto -> ScriptHash crypto
+policyID (PolicyID pid) = pid
+
+-- | A cryptocurrency value. This is a map from 'PolicyID's to a
+-- quantity of assets with this policy.
+--
+-- Operations on assets are usually implemented /pointwise/. That is,
+-- we apply the operation to the quantities for each asset in turn. So
+-- when we add two 'Value's the resulting 'Value' has, for each asset,
+-- the sum of the quantities of /that particular/ asset in the argument
+-- 'Value'. The effect of this is that the assets in the 'Value' are "independent",
+-- and are operated on separately.
+--
+-- Whenever we need to get the quantity of an asset in a 'Value' where there
+-- is no explicit quantity of that asset in the 'Value', then the quantity is
+-- taken to be zero.
+--
+-- See note [Assets] for more details.
+
+-- | Value type
+data Value crypto = Value
+  { val :: Map  (PolicyID crypto) (Map AssetID Quantity) }
+  deriving (Show, Generic)
+
+instance NoUnexpectedThunks (Value crypto)
+
+-- data ValueBSType = ValueBSType (Map ByteString (Map ByteString Quantity))
+--   deriving (Show, Eq, Generic)
+--
+-- instance NoUnexpectedThunks ValueBSType
+--
+-- -- | make a crypto-free Value type
+-- -- TODO this is a hack!
+-- toValBST :: Value crypto -> ValueBSType
+-- toValBST (Value v) = ValueBSType $ fromList $ fmap (\(cid, tkns) -> (pack $ show cid, tkns)) (toList v)
+
+-- | compact representation of Value
```

Comment:
What are the invariants? Is it legal to have a MixValue that has only Ada in it? What if there are other things but with 0 quantity?

Reply:
What are the rules about when this can be used? This needs a big note explaining what it's for and where it's supposed to be used.

Reply:
IMO the use of this is currently somewhat incoherent. I think we need to take one of two approaches:
- Just use `CompactValue` to define `Value`'s CBOR instances, don't export it.
- Use `CompactValue` everywhere we have a type that gets serialized, don't give `Value` CBOR instances to make sure we never use it accidentally in that way.

Or do we want to use `CompactValue` for faster calculation in some cases? I don't know, you need to write something about it!

(And if we're not using it for calculation, we should probably drop most of the operations on it like the `Semigroup` instance.)

Reply:
CompactValue now is built only by removing 0s and making a special case for Ada. It is needed for the UTxO